src/dune_engine/forker.ml

```diff
@@ -0,0 +1,203 @@
+open Import
+open Fiber.O
+module Dune_rpc = Dune_rpc_private
+
+module Decl : sig
+  val exec :
+    (Action_exec.input, Action_exec.Exec_result.t) Dune_rpc.Decl.request
+
+  val ready : (string, unit) Dune_rpc.Decl.request
+end = struct
+  module Conv = Dune_rpc_private.Conv
+  module Decl = Dune_rpc_private.Decl
+
+  module Exec = struct
+    let marshal () =
+      let to_ data = Marshal.from_string data 0 in
+      let from action = Marshal.to_string action [] in
+      Conv.iso Conv.string to_ from
+
+    let decl =
+      let v1 =
+        Decl.Request.make_current_gen ~req:(marshal ()) ~resp:(marshal ())
+          ~version:1
+      in
+      Decl.Request.make ~method_:"action/exec" ~generations:[ v1 ]
+  end
+
+  module Ready = struct
+    let decl =
+      let v1 =
+        Decl.Request.make_current_gen ~req:Conv.string ~resp:Conv.unit
+          ~version:1
+      in
+      Decl.Request.make ~method_:"action/ready" ~generations:[ v1 ]
+  end
+
+  let exec = Exec.decl
+
+  let ready = Ready.decl
+end
+
+module Client = Dune_rpc_client
+
+module Foreman = struct
+  type session = Session : _ Dune_rpc_server.Session.t -> session
+
+  type initialized =
+    { session : session
+    ; id : (module Stdune.Id.S)
+    }
+
+  type status =
+    | Awaiting_initialization of unit Fiber.Ivar.t
+    | Initialized of initialized
+    | Closed
+
+  let dyn_of_status =
+    let open Dyn in
+    function
+    | Awaiting_initialization _ -> variant "Awaiting_initialization" []
+    | Initialized _ -> variant "Initialized" []
+    | Closed -> variant "Closed" []
+
+  module Id = Stdune.Id.Make ()
+
+  type worker =
+    { pid : Pid.t
+    ; name : string
+    ; id : Id.t
+    ; mutable status : status
+    ; pool : Fiber.Pool.t
+    }
+
+  let await worker = Fiber.Pool.run worker.pool
+
+  let workers = Table.create (module String) 16
+
+  let _dyn_of_worker { pid; name; id; status; pool = _ } =
+    let open Dyn in
+    record
+      [ ("pid", Pid.to_dyn pid)
+      ; ("name", string name)
+      ; ("id", Id.to_dyn id)
+      ; ("status", dyn_of_status status)
+      ]
+
+  let register pid ~name =
```

Comment:
Passing in the pid seems mostly fine as we own the forker, but perhaps we should pass in something like `start: () -> Pid.t` to exclude edge case pid reuse errors?

Reply:
Good point. Let me give that a try.

Reply:
Wait, actually I'm not sure how adding `start ()` would help with pid reuse. To avoid pid reuse, we need the `Foreman.await` to just reap the pid and cleanup.